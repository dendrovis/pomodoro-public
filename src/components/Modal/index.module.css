.container {
  background-color: var(--tertiary-color-dark-translucent, grey);
  width: 100vw;
  max-width: calc(var(--max-width));
  min-height: 100vh;
}
