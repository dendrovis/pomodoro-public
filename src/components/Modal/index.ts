import Classes from './index.module.css';

type States = Properties & { elementContainer?: HTMLDialogElement };
type Properties = { id: string; content: Element; targetElementId: string; isOpen: boolean };
export default class Modal {
  #states: States;

  constructor(params: Properties) {
    this.#states = params;
  }

  isOpen(value: boolean) {
    this.#states.isOpen = value;
    if (this.#states.elementContainer) this.#states.elementContainer.open = this.#states.isOpen;
  }

  #createElementContainer() {
    const element = document.createElement('dialog') as HTMLDialogElement;
    element.id = this.#states.id;
    element.className = `${Classes.container}`;
    element.open = this.#states.isOpen;
    this.#states.elementContainer = element;
    return element;
  }

  #appendElement(elementSource: Element, elementDestination: Element) {
    elementDestination.appendChild(elementSource);
  }

  render() {
    const elementExist = document.querySelector(`#${this.#states.id}`);
    if (elementExist) throw new Error('element existed please try a new id');
    const element = document.querySelector(this.#states.targetElementId);
    if (!element) throw new Error('undefined element');
    const elementContainer = this.#createElementContainer();
    this.#appendElement(this.#states.content, elementContainer);
    this.#appendElement(elementContainer, element);
  }
}
