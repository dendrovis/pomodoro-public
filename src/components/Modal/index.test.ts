import { describe, it, expect, beforeAll, beforeEach } from 'vitest';
import Modal from '.';

beforeEach(() => {
  document.body.innerHTML = '';
});
beforeAll(() => {
  document.body.id = 'app';
});

describe('when enter id', () => {
  it('should render if id is distinct', () => {
    const element = document.createElement('div');
    const modal = new Modal({
      id: 'test',
      content: element,
      isOpen: false,
      targetElementId: '#app',
    });
    modal.render();
    expect(document.body.innerHTML).matchSnapshot();
  });
  it('should throw error if id exist', () => {
    const element = document.createElement('div');
    const modal = new Modal({
      id: 'test',
      content: element,
      isOpen: false,
      targetElementId: '#app',
    });
    modal.render();
    const modal_with_same_id = new Modal({
      id: 'test',
      content: element,
      isOpen: false,
      targetElementId: '#app',
    });
    expect(() => modal_with_same_id.render()).toThrowError();
  });
});

describe('when given content', () => {
  it('should render whatever is given to the content', () => {
    const element = document.createElement('div');
    element.innerHTML = `<div>test</div>`;
    const modal = new Modal({
      id: 'test',
      content: element,
      isOpen: false,
      targetElementId: '#app',
    });
    modal.render();
    expect(document.body.innerHTML).matchSnapshot();
  });
});

describe('when isOpen is set', () => {
  it('should display the content if true', () => {
    const element = document.createElement('div');
    const modal = new Modal({
      id: 'test',
      content: element,
      isOpen: true,
      targetElementId: '#app',
    });
    modal.render();
    expect(document.body.innerHTML).matchSnapshot();
  });
  it('should not display the content if false', () => {
    const element = document.createElement('div');
    const modal = new Modal({
      id: 'test',
      content: element,
      isOpen: false,
      targetElementId: '#app',
    });
    modal.render();
    expect(document.body.innerHTML).matchSnapshot();
  });
});

describe('when given targetElementId', () => {
  it('should render if targetElementId exist', () => {
    const element = document.createElement('div');
    const modal = new Modal({
      id: 'test',
      content: element,
      isOpen: false,
      targetElementId: '#app',
    });
    modal.render();
    expect(document.body.innerHTML).matchSnapshot();
  });
  it('should throw error if does not exist', () => {
    const element = document.createElement('div');
    const modal = new Modal({
      id: 'test',
      content: element,
      isOpen: false,
      targetElementId: 'app',
    });

    expect(() => modal.render()).toThrowError();
  });
});

describe('when toggle is set', () => {
  it('should display the content if toggle when isOpen is false', () => {
    const element = document.createElement('div');
    const modal = new Modal({
      id: 'test',
      content: element,
      isOpen: false,
      targetElementId: '#app',
    });
    modal.render();
    modal.isOpen(false);

    expect(document.body.innerHTML).toMatchSnapshot();
  });
  it('should not display the content if toggle when isOpen is true', () => {
    const element = document.createElement('div');
    const modal = new Modal({
      id: 'test',
      content: element,
      isOpen: true,
      targetElementId: '#app',
    });
    modal.render();
    modal.isOpen(true);
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});
