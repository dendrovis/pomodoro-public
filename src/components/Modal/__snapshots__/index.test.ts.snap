// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`when enter id > should render if id is distinct 1`] = `"<dialog id="test" class="_container_36b8ff"><div></div></dialog>"`;

exports[`when given content > should render whatever is given to the content 1`] = `"<dialog id="test" class="_container_36b8ff"><div><div>test</div></div></dialog>"`;

exports[`when given targetElementId > should render if targetElementId exist 1`] = `"<dialog id="test" class="_container_36b8ff"><div></div></dialog>"`;

exports[`when isOpen is set > should display the content if true 1`] = `"<dialog id="test" class="_container_36b8ff" open=""><div></div></dialog>"`;

exports[`when isOpen is set > should not display the content if false 1`] = `"<dialog id="test" class="_container_36b8ff"><div></div></dialog>"`;

exports[`when toggle is set > should display the content if toggle when isOpen is false 1`] = `"<dialog id="test" class="_container_36b8ff"><div></div></dialog>"`;

exports[`when toggle is set > should not display the content if toggle when isOpen is true 1`] = `"<dialog id="test" class="_container_36b8ff" open=""><div></div></dialog>"`;
